src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -381,8 +385,25 @@ private async Task RefreshIndividualKeyValues()
                                 break;
                             }
 
-                            hasChanged = true;
-                            ProcessChanges(Enumerable.Repeat(keyValueChange, 1));
+                            if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)

```

Comment:
>keyValueChange.ChangeType == KeyValueChangeType.Deleted [](start = 32, length = 55)

Don't we need to do this before breaking from the loop? Seems like otherwise we will be storing an outdated kv.

Reply:
I thought more about this. If we change the cached values before the `break` statement, and lets say `refreshAll` operation fails due to some `RequestFailedException`, we have now cached the latest value of a sentinel key without updating application configuration. On the next call to `TryRefreshAsync`, since we already have the latest value of sentinel key, we will not be able to observe any changes to the sentinel key.
Instead of changing the logic here, I think it would be better to clear the `_watchedSettings` cache every time `LoadKeyValuesRegisteredForRefresh` is called (since we are always going to fetch the latest values of refresh registered keys anyway).

Reply:
Won't we have that problem because of this line already? https://github.com/Azure/AppConfiguration-DotnetProvider/blob/2ab6bd32470523d918cee545bc2e135d2b20fb20/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs#L377

Reply:
Yes, the next refresh operation will be triggered only after the cache expires again. This is our behavior even today. If the first refresh failed, wait for `CacheExpires` time before triggering another refresh.

Reply:
I remember Abhilash and I discussed this in the past and ultimately this is a known behavior.

Reply:
Talked with Avani. It may look like the update being here could cause us to maintain stale key-values in the watchedsettings list, but actually the refreshAll operation will cause the updated values to be used.

Secondly, if we were to move it above the refreshAll + break, then we would never trigger a refresh again on a failure. Right now, if there is a failure in reload all, we would trigger the refresh again when the watcher's cache expires.